backend/aopp.go

```diff
@@ -0,0 +1,303 @@
+	ErrorCode ErrorCode `json:"errorCode"`
+	// Accounts is the list of accounts the user can choose from. Only applies if State == aoppStateChoosingAccount.
+	Accounts []account `json:"accounts"`
+	// Address that will be delivered to the requesting party via the callback. Only applies if State == aoppStateSigning.
+	Address string `json:"address"`
+	// CallbackHost contains the host of the AOPP callback URL. Available for all states except
+	// aoppStateInactive.
+	CallbackHost string `json:"callbackHost"`
+	// coinCode is the requested asset. Available for all states except aoppStateInactive.
+	coinCode coinpkg.Code
+	// message is the requested message to be signed. Available for all states except
+	// aoppStateInactive.
+	message string
+	// callback is the AOPP callback param. Available for all states except aoppStateInactive.
+	callback string
+}
+
+// AOPP returns the current AOPP state.
+func (backend *Backend) AOPP() AOPP {
+	defer backend.accountsAndKeystoreLock.RLock()()
+	return backend.aopp
+}
+
+// notifyAOPP sends the aopp state to the frontend. `accountsAndKeystoreLock` must be held when
+// calling this function.
+func (backend *Backend) notifyAOPP() {
+	backend.Notify(observable.Event{
+		Subject: "aopp",
+		Action:  action.Replace,
+		Object:  backend.aopp,
+	})
+}
+
+// AOPPCancel resets the aopp state.
+func (backend *Backend) AOPPCancel() {
+	defer backend.accountsAndKeystoreLock.Lock()()
+	backend.aopp = AOPP{State: aoppStateInactive}
+	backend.notifyAOPP()
+}
+
+// aoppSetError pushes an error to the frontend to display. `accountsAndKeystoreLock` must be held
+// when calling this function.
+func (backend *Backend) aoppSetError(err ErrorCode) {
+	backend.aopp.State = aoppStateError
+	backend.aopp.ErrorCode = err
+	backend.notifyAOPP()
+}
+
+// aoppKeystoreRegistered must be called after a keystore is available, to display a list of
+// accounts to choose from. `accountsAndKeystoreLock` must be held when calling this function.
+func (backend *Backend) aoppKeystoreRegistered() {
+	if backend.aopp.State != aoppStateAwaitingKeystore {
+		return
+	}
+	if !backend.keystore.CanSignMessage(backend.aopp.coinCode) {
+		backend.aoppSetError(errAoppCannotSign)
+		return
+	}
+	var accounts []account
+	for _, acct := range backend.accounts {
+		if !acct.Config().Active {
+			continue
+		}
+		if acct.Coin().Code() != backend.aopp.coinCode {
+			continue
+		}
+		accounts = append(accounts, account{
+			Name: acct.Config().Name,
+			Code: acct.Config().Code,
+		})
+	}
+
+	if len(accounts) == 0 {
+		backend.aoppSetError(errAoppNoAccounts)
+		return
+	}
+
+	backend.aopp.Accounts = accounts
+	backend.aopp.State = aoppStateChoosingAccount
+	backend.notifyAOPP()
+}
+
+// handleAOPP handles an AOPP (Address Ownership Proof Protocol) request. See https://aopp.group/.
+func (backend *Backend) handleAOPP(uri *url.URL) {
+	defer backend.accountsAndKeystoreLock.Lock()()
+
+	log := backend.log.WithField("aopp-uri", uri.String())
+
+	q := uri.Query()
```

Comment:
since `uri` arg is a pointer, this can lead to null pointer dereference.

Reply:
isn't that just how Go works? half the stuff is passed by pointer to avoid copies :(

Reply:
Passing by value now.

Reply:
I wouldn't worry about memory optimization here. There are lots of much bigger allocations in the app. But if you do worry, here's an article: https://medium.com/a-journey-with-go/go-should-i-use-a-pointer-instead-of-a-copy-of-my-struct-44b43b104963

Takeaway:

> Using a copy of the struct instead of a pointer here is 8 times faster.